login/src/main/java/com/fightpandemics/login/ui/profile/CompleteProfileFragment.kt

```diff
@@ -113,15 +176,125 @@ class CompleteProfileFragment : Fragment() {
+        fragmentCompleteProfileBinding.root.etAddress.doAfterTextChanged {
+            if (fragmentCompleteProfileBinding.root.etAddress.hasFocus()) {
+                it.toString().also { inputLocation ->
+                    handleAutocompleteVisibility(inputLocation)
+                    searchLocationPredictions(inputLocation)
+                }
+                // reset location selected since we are searching for a new location
+                loginViewModel.resetLocation()
+            }
+        }
+    }
+
+    private fun searchLocationPredictions(inputLocation: String) {
+        lifecycleScope.launchWhenStarted {
+            loginViewModel.searchQuery.value = inputLocation
+            loginViewModel.searchLocationState.collect { preditions ->
+                adapter.placesNames = preditions
+            }
+        }
+    }
+
+    private fun onSelectedLocation(selectedLocation: String) {
+        // display selected location in editText
+        fragmentCompleteProfileBinding.root.etAddress.setText(selectedLocation)
+        // save location selected in liveData view model to prevent not-selected input from being accepted
+        loginViewModel.locationSelected.value = selectedLocation
+        removeFocusFromLocationEditText()
+    }
+
+    private fun handleAutocompleteVisibility(locationQuery: String, hasFocus: Boolean = true) {
+        if (locationQuery.isBlank() || locationQuery.length < LENGTH_TO_SHOW_SUGGESTIONS || !hasFocus) {
+            fragmentCompleteProfileBinding.root.auto_complete_locations_recycler_view.visibility =
+                View.GONE
+            fragmentCompleteProfileBinding.root.item_line_divider1.visibility = View.GONE
+        } else {
+            fragmentCompleteProfileBinding.root.auto_complete_locations_recycler_view.visibility =
+                View.VISIBLE
+            fragmentCompleteProfileBinding.root.item_line_divider1.visibility = View.VISIBLE
+        }
+    }
+
+    private fun handleLocationSelectedText(hasFocus: Boolean) {
+        if (!hasFocus && loginViewModel.locationSelected.value.isNullOrBlank()) {
+            fragmentCompleteProfileBinding.root.etAddress.text?.clear()
+        }
+    }
+
+    // This removes focus and hides keyboard
+    private fun removeFocusFromLocationEditText() {
+        fragmentCompleteProfileBinding.root.tilLocation.clearFocus()
+        dismissKeyboard(fragmentCompleteProfileBinding.root.tilLocation)
+    }
+
+    private fun bindLoading(isLoading: Boolean) {
+        fragmentCompleteProfileBinding.root.progressBar.visibility =
+            if (isLoading) View.VISIBLE else View.GONE
+    }
+
+    override fun onAutocompleteLocationClick(locationSelected: Prediction) {
+        onSelectedLocation(locationSelected.description)
+        loginViewModel.getDetails(locationSelected.place_id)
+    }
+
     companion object {
         const val USER_PROFILE = "userProfile"
+        const val LENGTH_TO_SHOW_SUGGESTIONS = 3
```

Comment:
Just for my own understanding, what would the difference be if this was a normal `object`?

Reply:
Companion objects are static and normal objects are Singleton. 

static means the attribute or method belongs to the class and not to the object instance.